beps/bep_0003.rst

```diff
@@ -87,70 +92,153 @@ announce
+  messages. It must be a power of two and at least 16KiB.
+  
+  Files are mapped into this piece address space so that each non-empty file starts
+  at a piece boundary and occurs in the same order as in the file tree.
+  The last piece of each file may be shorter than the specified piece length.
+  
+``meta version``
+  an integer value, set to 1 to indicate compatibility with the current revision of this
+  specification. Future revisions will only increment this value to indicate an incompatible
+  change has been made, for example that hash algorithms were changed due to newly discovered
+  vulnerabilities. Implementations must check this field first and indicate that a torrent
+  is of a newer version than they can handle before performing other validations which may
+  result in more general messages about invalid files.
+
+``file tree``
+  A tree of dictionaries where dictionary keys represent utf-8 encoded path elements
+  Entries with zero-length keys describe the properties of the composed path at that point
+  The ``file tree`` root dictionary itself must not be a file, i.e. it must not contain
+  a zero-length key with a dictionary containing a ``length`` key.
+
+
+File tree layout
+----------------
+
+Example:
+
+.. parsed-literal::
+
+    {
+      info: {
+        file tree: {
+          dir1: {
+            dir2: {
+              fileA.txt: {
+                "": {
+                  length: *<length of file in bytes (integer)>*,
+                  pieces root: *<optional, merkle tree root (string)>*,
+                  ...
+                }
+              },
+              fileB.txt: {
+                "": {
+                  ...
+                }
+              }
+            },
+            dir3: {
+              ...
+            }
+          }
+        }
+      }
+    }
+    
+Bencoded for fileA only::
+
+    d4:infod9:file treed4:dir1d4:dir2d9:fileA.txtd0:d5:lengthi1024e11:pieces root32:aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaeeeeeee
 
-``piece length`` maps to the number of bytes in each piece
-the file is split into. For the purposes of transfer, files are
-split into fixed-size pieces which are all the same length except for
-possibly the last one which may be truncated. ``piece
-length`` is almost always a power of two, most commonly 2 18 =
-256 K (BitTorrent prior to version 3.2 uses 2 20 = 1 M as
-default).
 
-``pieces`` maps to a string whose length is a multiple of
-20. It is to be subdivided into strings of length 20, each of which is
-the SHA1 hash of the piece at the corresponding index.
+``length``
+  Length of the file in bytes. Presence of this field indicates
+  that the dictionary describes a file, not a directory. Which means
+  it must not have any sibling entries.
+  
+``pieces root``
+  For non-empty files this is the the root hash of a merkle tree
+  with a branching factor of 2, constructed from 16KiB blocks of the file.
+  If necessary the last block of the file must be zero-padded to 16KiB.
```

Comment:
Hrm, I guess there's a point here. But I don't know if we should isolate the piece layer from the file representation and just have it hash full, padded pieces all the time or let the file sizes affect the piece hashing.